src/multi.rs

```diff
@@ -83,48 +69,79 @@ impl<'lua, T: FromLua<'lua>> FromLuaMulti<'lua> for Variadic<T> {
+            #[allow(unused_mut)]
+            #[allow(non_snake_case)]
+            fn to_lua_multi(self, lua: &'lua Lua) -> Result<MultiValue<'lua>> {
+                let ($($name,)* $last,) = self;
+
+                let mut results = $last.to_lua_multi(lua)?;
+                push_reverse!(results, $($name.to_lua(lua)?,)*);
+                Ok(results)
+            }
+        }
 
-impl<'lua, T: ToLuaMulti<'lua>> ToLuaMulti<'lua> for HCons<T, HNil> {
-    fn to_lua_multi(self, lua: &'lua Lua) -> Result<MultiValue<'lua>> {
-        self.0.to_lua_multi(lua)
-    }
+        impl<'lua, $($name,)* $last> FromLuaMulti<'lua> for ($($name,)* $last,)
+            where $($name: FromLua<'lua>,)*
+                  $last: FromLuaMulti<'lua>
+        {
+            #[allow(unused_mut)]
+            #[allow(non_snake_case)]
+            fn from_lua_multi(mut values: MultiValue<'lua>, lua: &'lua Lua) -> Result<Self> {
+                $(let $name = values.pop_front().unwrap_or(Nil);)*
+                let $last = FromLuaMulti::from_lua_multi(values, lua)?;
+                Ok(($(FromLua::from_lua($name, lua)?,)* $last,))
+            }
+        }
+    );
 }
 
-impl<'lua, T: FromLuaMulti<'lua>> FromLuaMulti<'lua> for HCons<T, HNil> {
-    fn from_lua_multi(values: MultiValue<'lua>, lua: &'lua Lua) -> Result<Self> {
-        Ok(HCons(T::from_lua_multi(values, lua)?, HNil))
-    }
-}
+macro_rules! push_reverse {
+    ($multi_value:expr, $first:expr, $($rest:expr,)*) => (
+        push_reverse!($multi_value, $($rest,)*);
+        $multi_value.push_front($first);
+    );
 
-impl<'lua, H: ToLua<'lua>, A, B> ToLuaMulti<'lua> for HCons<H, HCons<A, B>>
-where
-    HCons<A, B>: ToLuaMulti<'lua>,
-{
-    fn to_lua_multi(self, lua: &'lua Lua) -> Result<MultiValue<'lua>> {
-        let mut results = self.1.to_lua_multi(lua)?;
-        results.push_front(self.0.to_lua(lua)?);
-        Ok(results)
-    }
-}
+    ($multi_value:expr, $first:expr) => (
+        $multi_value.push_front($first);
+    );
 
-impl<'lua, H: FromLua<'lua>, A, B> FromLuaMulti<'lua> for HCons<H, HCons<A, B>>
-where
-    HCons<A, B>: FromLuaMulti<'lua>,
-{
-    fn from_lua_multi(mut values: MultiValue<'lua>, lua: &'lua Lua) -> Result<Self> {
-        let val = H::from_lua(values.pop_front().unwrap_or(Nil), lua)?;
-        let res = HCons::<A, B>::from_lua_multi(values, lua)?;
-        Ok(HCons(val, res))
-    }
+    ($multi_value:expr,) => ();
 }
+
+impl_tuple! { }
+impl_tuple! { A }
+impl_tuple! { A B }
+impl_tuple! { A B C }
+impl_tuple! { A B C D }
+impl_tuple! { A B C D E }
+impl_tuple! { A B C D E F }
+impl_tuple! { A B C D E F G }
+impl_tuple! { A B C D E F G H }
+impl_tuple! { A B C D E F G H I }
+impl_tuple! { A B C D E F G H I J }
+impl_tuple! { A B C D E F G H I J K }
+impl_tuple! { A B C D E F G H I J K L }
+impl_tuple! { A B C D E F G H I J K L M }
+impl_tuple! { A B C D E F G H I J K L M N }
+impl_tuple! { A B C D E F G H I J K L M N O }
+impl_tuple! { A B C D E F G H I J K L M N O P }
```

Comment:
I'm wondering if these impact compile times at all. Only going up to 8-tuples would probably still be enough (actually, a function taking 8 unnamed arguments is already far too confusing for my taste).

The stdlib seems to go up to 12, so that's another option for consistency.